src/async-interleave-generator.mjs

```diff
@@ -0,0 +1,44 @@
+import { ensureAsyncIterable } from './internal/async-iterable'
+import raceTo from './internal/race-to'
+import map from './map'
+
+class AsyncInterleaveBuffer {
+  constructor (iterator) {
+    this._iterator = iterator
+    this._nextPromise = iterator.next()
+  }
+
+  async take () {
+    const next = await this._nextPromise
+    if (next.done) {
+      return undefined
+    }
+
+    this._nextPromise = this._iterator.next()
+    return next.value
+  }
+
+  async canTake () {
+    return !(await this._nextPromise).done
+  }
+}
+
+export default function interleaveGenerator (generatorFn) {
+  return (...iterables) => {
+    return (async function * () {
+      const buffers = iterables.map(iterable =>
+         new AsyncInterleaveBuffer(ensureAsyncIterable(iterable)[Symbol.asyncIterator]())
+      )
+      const canTakeAny = () => raceTo(Boolean, false, map(async buffer => (await buffer.canTake()) ? buffer : null, buffers))
```

Comment:
I see this:

```javascript
map(async buffer => (await buffer.canTake()) ? buffer : null, buffers)
```

Why don't you just use `buffers.map`?

Reply:
I also have another question:

So `raceTo` returns a **promise** that resolves an object (`Promise<AsyncInterleaveBuffer>`) that has 2 methods (`canTake` and `take`) that return promise again?

This would require user to do at least 3 `await` in callback function, one for the `AsyncInterleaveBuffer`, one for `canTake` and one for `take`.

Why don't you pass a wrapper instead? i.e.

```javascript
// your old class
class AsyncInterleaveBuffer { /* ... */ }

// the wrapper
const createCanTakeAny = (...args) => () => {
  const bufferPromise = raceTo(...args)
  return {
    async canTake () {
      return (await bufferPromise).canTake()
    },
    async take () {
      return (await bufferPromise).take()
    }
  }
}
```

This way, user wouldn't have to `await` to get `AsyncInterleaveGenerator`

Reply:
Why would I use buffers.map? It creates an array I don't need to allocate and it would create that array for every item in every iterable.

Reply:
To your second question, that's right. I don't think the overhead of writing await a few times is worth paying any significant price to fix, and I think what you are proposing has cost. It has perf cost, and it is likely to either be different from the sync form of the API or unnecessarily complicate it.

Reply:
> It has perf cost

It's always going to be, either from the code of `iter-tools`, or from the user who is going to use `await` to turn the promise into something useful.

> and it is likely to either be different from the sync form of the API

It already different from the sync form (e.g. `buffer.take()` return a promise that resolves a value instead of the value directly).